From 8fcfcbd926cf1b82e032743e1e36098a4971106e Mon Sep 17 00:00:00 2001
From: Jes Sorensen <Jes.Sorensen@redhat.com>
Date: Mon, 18 Aug 2014 13:27:39 -0400
Subject: [md] raid6: avoid data corruption during recovery of double-degraded RAID6

Message-id: <1408368459-7970-2-git-send-email-Jes.Sorensen@redhat.com>
Patchwork-id: 87855
O-Subject: [RHEL7.1 PATCH 1/1] md/raid6: avoid data corruption during recovery of double-degraded RAID6
Bugzilla: 1130905
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Doug Ledford <dledford@redhat.com>

commit 9c4bdf697c39805078392d5ddbbba5ae5680e0dd
From: NeilBrown <neilb@suse.de>

During recovery of a double-degraded RAID6 it is possible for
some blocks not to be recovered properly, leading to corruption.

If a write happens to one block in a stripe that would be written to a
missing device, and at the same time that stripe is recovering data
to the other missing device, then that recovered data may not be written.

This patch skips, in the double-degraded case, an optimisation that is
only safe for single-degraded arrays.

Bug was introduced in 2.6.32 and fix is suitable for any kernel since
then.  In an older kernel with separate handle_stripe5() and
handle_stripe6() functions that patch must change handle_stripe6().

Cc: stable@vger.kernel.org (2.6.32+)
Fixes: 6c0069c0ae9659e3a91b68eaed06a5c6c37f45c8
Cc: Yuri Tikhonov <yur@emcraft.com>
Cc: Dan Williams <dan.j.williams@intel.com>
Reported-by: "Manibalan P" <pmanibalan@amiindia.co.in>
Resolves: https://bugzilla.redhat.com/show_bug.cgi?id=1090423
Signed-off-by: NeilBrown <neilb@suse.de>
(cherry-picked from f94e37dce722ec7b6666fd04be357f422daa02b5)
(https://lkml.org/lkml/2014/8/18/17)
Signed-off-by: Jes Sorensen <Jes.Sorensen@redhat.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/raid5.c b/drivers/md/raid5.c
index d825059..fe42c3b 100644
--- a/drivers/md/raid5.c
+++ b/drivers/md/raid5.c
@@ -3560,6 +3560,8 @@ static void handle_stripe(struct stripe_head *sh)
     set_bit(R5_Wantwrite, &dev->flags);
     if (prexor)
      continue;
+    if (s.failed > 1)
+     continue;
     if (!test_bit(R5_Insync, &dev->flags) ||
         ((i == sh->pd_idx || i == sh->qd_idx)  &&
          s.failed == 0))
-- 
1.7.1